### PR TITLE
feat: 0x07-smod

### LINKED ISSUE #94: feat: 0x07 SMOD
<div class="text-sm px-4 md:px-8 py-8 bg-indigo-50 dark:bg-black-600"><table class="table-auto mb-6 bg-indigo-100 dark:bg-black-500 rounded font-medium"><thead><tr class="text-gray-500 uppercase text-xs tracking-wide"><td class="pt-3 px-4">Since</td><td class="pt-3 px-4">Group</td></tr></thead><tbody><tr><td class="pb-3 px-4">Frontier</td><td class="pb-3 px-4">Stop and Arithmetic Operations</td></tr></tbody></table><div class="flex flex-col lg:flex-row"><div class=""><p class="leading-5 mb-4"><em>Index 1 is top of the stack. See <a class="underline" href="/#60">PUSH</a>.</em></p><h2 class="text-base font-semibold my-4">Notes</h2><p class="leading-5 mb-4">All values are treated as two’s complement signed 256-bit integers. Note the overflow semantic when −2<sup>255</sup> is negated.</p><h2 class="text-base font-semibold my-4">Stack input</h2><ol class="list-decimal mb-4"><li class="ml-6"><code>a</code>: integer numerator.</li><li class="ml-6"><code>b</code>: integer denominator.</li></ol><h2 class="text-base font-semibold my-4">Stack output</h2><ol class="list-decimal mb-4"><li class="ml-6"><code>a % b</code>: integer result of the signed integer modulo. If the denominator is 0, the result will be 0.</li></ol><h2 class="text-base font-semibold my-4">Examples</h2><table class="table-auto mb-4"><thead><tr><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all"></th><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all border">Input</th><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all border">Output</th></tr></thead><tbody><tr><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>1</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>10</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>1</code></td></tr><tr><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>2</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>3</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"></td></tr></tbody></table><table class="table-auto mb-4"><thead><tr><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all"></th><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all border">Input</th><th class="py-1 px-2 border-indigo-200 dark:border-black-400 text-gray-800 dark:text-gray-400 text-tiny font-medium break-all border">Output</th></tr></thead><tbody><tr><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>1</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF8</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE</code></td></tr><tr><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>2</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"><code>0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD</code></td><td class="py-1 px-2 border-indigo-200 dark:border-black-400 text-tiny font-normal break-all border"></td></tr></tbody></table><p class="leading-5 mb-4"><a class="underline" href="https://evm.codes/playground?unit=Wei&amp;codeType=Mnemonic&amp;code=%27y1s3s10tzzy2wrDwr8t%27~uuuz%5Cny%2F%2F%20Example%20wv32%200xr~vzPUSHuFFFtzSMODsv1%20r~~~%01rstuvwyz~_">Reproduce in playground</a>.</p><h2 class="text-base font-semibold my-4">Error cases</h2><p class="leading-5 mb-4">The state changes done by the current context are <a class="underline" href="/?fork=shanghai#FD">reverted</a> in those cases:</p><ul class="list-disc mb-2"><li class="ml-6">Not enough gas.</li><li class="ml-6">Not enough values on the stack.</li></ul></div></div></div>

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: Closes #94 

## What is the new behavior?

- Implements 0x07-SMOD opcode
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/tests/test_instructions/test_stop_and_arithmetic_operations.cairo [hunk 1]
```diff
@@ -222,13 +222,68 @@ fn test_exec_mod_by_zero() {
...
+    ctx.exec_smod();
+
+    // Then
+    assert(ctx.stack.len() == 1, 'stack should have one element');
+    assert(
+        ctx
+            .stack
+            .peek()
+            .unwrap() == 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE,
+        'stack top should be -8%-3 = -1'
```
Comment (Eikix) [inline src/tests/test_instructions/test_stop_and_arithmetic_operations.cairo:266]:
```suggestion
        'stack top should be -8%-3 = -2'
```

File: src/instructions/stop_and_arithmetic_operations.cairo [hunk 2]
```diff
@@ -130,7 +130,22 @@ impl StopAndArithmeticOperations of StopAndArithmeticOperationsTrait {
     /// a % b: integer result of the signed integer modulo. If the denominator is 0, the result will be 0.
     /// All values are treated as two’s complement signed 256-bit integers. Note the overflow semantic when −2^255 is negated.
     /// # Specification: https://www.evm.codes/#07?fork=shanghai
-    fn exec_smod(ref self: ExecutionContext) {}
+    fn exec_smod(ref self: ExecutionContext) {
+        let popped = self.stack.pop_n(2);
+
+        let a: u256 = *popped[0];
```
Comment (Eikix) [inline src/instructions/stop_and_arithmetic_operations.cairo:136]: random idea but should we name the variables explicit names?

`a` and `denominator` for instance

Comment (Eikix) [reply src/instructions/stop_and_arithmetic_operations.cairo:136]: if you agree we can create an issue now and refactor some opcodes later

**Review by Eikix — COMMENTED** [review]
almost gtm

File: src/instructions/stop_and_arithmetic_operations.cairo (hunk 2 above)
Comment (enitrat) [reply src/instructions/stop_and_arithmetic_operations.cairo:136]: in evm.codes they're named a and b, but we can

**Review by Eikix — APPROVED** [review]
lgtm gg

**Pull request merged**